From 21d663c1e31abe815bcb164f41953a801bad42bf Mon Sep 17 00:00:00 2001
From: Niels de Vos <ndevos@redhat.com>
Date: Fri, 30 Jun 2017 13:24:01 +0200
Subject: [PATCH] build: do not require 'git' to find the version

While looking into packaging gluster-block for Fedora, I noticed that
'git' is used to determine the version. The downloads from GitHub are
not archives that have been created with 'make dist', and hence
./autogen.sh is still needed to be run.

./autogen.sh generates all the needed bits, including the ./configure
script that tries to figure out the version. If ./configure runs in a
non-git directory, no version can be found.

In order to solve this, I'm adding a VERSION file in the root of the
project. If this file exists, it is used to fetch the version, instead
of trying to run git. The basic usage for RPM packaging in distributions
can then be:

  echo %{version} > VERSION
  ./autogen.sh
  %configure

Note that the VERSION file is not needed for building from a git
repository. However, this file can get packaged in the 'make dist'
tarball so that building from the tarball does not require git.

Change-Id: Ied378c7071ee4a108a1e946dccbb7f223b7aeb9f
Updates: #25
Signed-off-by: Niels de Vos <ndevos@redhat.com>
---
 .gitignore   | 1 +
 Makefile.am  | 7 +++++--
 configure.ac | 2 +-
 3 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/.gitignore b/.gitignore
index d66bf2f..45c9efa 100644
--- a/.gitignore
+++ b/.gitignore
@@ -23,6 +23,7 @@ stamp-h1
 *.spec
 *.service
 *.tar.gz
+VERSION
 
 cscope.*
 tags
diff --git a/Makefile.am b/Makefile.am
index 8e63b2c..7e5018b 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -40,9 +40,12 @@ gitclean: clean distclean
 		ltmain.sh missing stamp-h1
 	rm -rf autom4te.cache ar-lib *.tar.gz m4
 
-dist-hook: gen-ChangeLog
+dist-hook: gen-ChangeLog gen-VERSION
 
-.PHONY: gen-ChangeLog
+.PHONY: gen-ChangeLog gen-VERSION
 
 gen-ChangeLog:
 	(cd $(srcdir) && git diff && echo ===== git log ==== && git log) > $(distdir)/ChangeLog
+
+gen-VERSION:
+	(cd $(srcdir) && git describe --always --tags | cut -f2 -d'v' | cut -f1 -d'-') > $(distdir)/VERSION
diff --git a/configure.ac b/configure.ac
index 77152b5..56767a6 100644
--- a/configure.ac
+++ b/configure.ac
@@ -9,7 +9,7 @@ dnl  cases as published by the Free Software Foundation.
 
 AC_PREREQ([2.69])
 AC_INIT([gluster-block],
-        m4_esyscmd(echo -n $(git describe --always --tags | cut -f2 -d'v' | cut -f1 -d'-')),
+        m4_esyscmd(echo -n $(cat VERSION 2>/dev/null || git describe --always --tags | cut -f2 -d'v' | cut -f1 -d'-')),
         [pkalever@redhat.com],,
         [https://github.com/gluster/gluster-block.git])
 
-- 
2.9.3

